File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.py
from collections import deque

def solution(x, y, n):
    answer = 1000001
    q = deque()
    q.append((y, 0))
    while q:
        num, cnt = q.popleft()
        if num == x:
            answer = min(answer, cnt)
            break

        if num % 3 == 0:
            q.append((num // 3, cnt+1))
            
        if num % 2 == 0:
            q.append((num // 2, cnt+1))
            
        if num - n >= x:
            q.append((num - n, cnt+1))
            
    if answer == 1000001:
        answer = -1
    
    return answer

File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.java
import java.util.Queue;
import java.util.LinkedList;
class Solution {
    public int solution(int x, int y, int n) {
        int answer = 1000001;
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[] {y, 0});
        while (!q.isEmpty()) {
            int[] tmp = q.poll();
            
            if (tmp[0] == x) {
                answer = Math.min(answer, tmp[1]);
                break;
            }
            
            if (tmp[0] % 3 == 0) {
                q.offer(new int[] {tmp[0] / 3, tmp[1] + 1});
            }
            
            if (tmp[0] % 2 == 0) {
                q.offer(new int[] {tmp[0] / 2, tmp[1] + 1});
            }
            
            if (tmp[0] - n >= 0) {
                q.offer(new int[] {tmp[0] - n, tmp[1] + 1});
            }
        }
        
        if (answer == 1000001) {
            answer = -1;
        }
        return answer;
    }
}